### PR TITLE
Features/lock

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
Support lock service
* what changes does this pull request make?
Support lock service
* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
No

### CODE REVIEW

File: xline/src/storage/kvwatcher.rs [hunk 1]
```diff
@@ -203,10 +202,9 @@ pub(crate) trait KvWatcherOps {
     fn cancel(&self, watcher: &Watcher) -> i64;
 }
 
-#[async_trait::async_trait]
 impl KvWatcherOps for KvWatcher {
     /// Create a watch to KV store
-    async fn watch(
+    fn watch(
```
Comment (Phoenix500526) [inline xline/src/storage/kvwatcher.rs:207]: Why should we remove `async` here?

File: xline/src/server/kv_server.rs [hunk 2]
```diff
@@ -433,7 +433,7 @@ impl Kv for KvServer {
     ) -> Result<tonic::Response<TxnResponse>, tonic::Status> {
         debug!("Receive TxnRequest {:?}", request);
         Self::check_txn_request(request.get_ref())?;
-        let is_fast_path = true;
+        let is_fast_path = false; // lock need revision of txn
         let (cmd_res, sync_res) = self.propose(request, is_fast_path).await?;
```
Comment (markcty) [inline xline/src/server/kv_server.rs:438]: If we use slow path here, should `sync_res` always be Some(...)? Then we should at least give a warning if it's None. I think a better way is to seperate self.propose into two functions. E.g., propose_slow will always `SyncResponse`, while propose_fast will return `Option<SyncResponse>`

File: xline/src/server/xline_server.rs [hunk 3]
```diff
@@ -51,6 +52,8 @@ pub struct XlineServer {
     lease_storage: Arc<LeaseStore>,
     /// Consensus client
     client: Arc<Client<Command>>,
+    /// Xline client
+    xline_client: Arc<Mutex<crate::client::Client>>,
```
Comment (markcty) [inline xline/src/server/xline_server.rs:56]: Is it necessary to put xlien_client in a Mutex? I see no mutable usage. Maybe `Arc<Client>` is enough?

File: xline/src/client/mod.rs [hunk 4]
```diff
@@ -238,15 +249,33 @@ impl Client {
         Ok(response.into())
     }
 
+    /// Gets an kv client.
+    #[inline]
+    pub fn kv_client(&self) -> KvClient {
```
Comment (markcty) [inline xline/src/client/mod.rs:254]: Should we expose these sub clients? We have already exposed more high-level interface such as put, range, delete. I guess the kv_client should be purely internal.

File: xline/src/server/lock_server.rs [hunk 5]
```diff
@@ -26,54 +40,218 @@ pub(crate) struct LockServer {
...
+            })?;
+            Ok((cmd_res, Some(sync_res)))
+        }
     }
 
-    /// Generate propose id
-    fn generate_propose_id(&self) -> ProposeId {
-        ProposeId::new(format!("{}-{}", self.name, Uuid::new_v4()))
+    /// Crate txn for try acquire lock
+    fn try_acquire_txn(prefix: &str, lease_id: i64) -> TxnRequest {
```
Comment (Phoenix500526) [inline xline/src/server/lock_server.rs:128]: Why do we name this function as `try_acquire_txn`? The prefix `try_` looks like it should return a `Result`. In addition, this function only constructs a `txn` object. It should not fail.

**Review by markcty — CHANGES_REQUESTED** [review]

File: xline/src/client/mod.rs (hunk 4 above)
Comment (markcty) [reply xline/src/client/mod.rs:254]: So do following sub clients.

File: xline/tests/lock_test.rs [hunk 6]
```diff
@@ -0,0 +1,33 @@
+mod common;
+
+use std::{error::Error, time::Duration};
+
+use common::Cluster;
+use tokio::time;
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 10)]
+async fn test_lock() -> Result<(), Box<dyn Error>> {
```
Comment (Phoenix500526) [inline xline/tests/lock_test.rs:9]: Please add a test to cover the deadlock case to ensure that the lease expired will release the lock

File: xline/src/client/mod.rs [hunk 7]
```diff
@@ -238,15 +249,33 @@ impl Client {
         Ok(response.into())
     }
 
+    /// Gets an kv client.
```
Comment (Phoenix500526) [inline xline/src/client/mod.rs:252]: Typo. It should be "Gets a kv client"

File: xline/src/client/mod.rs [hunk 8]
```diff
@@ -238,15 +249,33 @@ impl Client {
...
+    }
+
     /// Gets an auth client.
     #[inline]
-    pub fn auth_client(&mut self) -> AuthClient {
+    pub fn auth_client(&self) -> AuthClient {
         self.etcd_client.auth_client()
     }
 
     /// Gets an watch client.
```
Comment (Phoenix500526) [inline xline/src/client/mod.rs:264]: Same above

File: xline/src/client/mod.rs [hunk 9]
```diff
@@ -238,15 +249,33 @@ impl Client {
...
     #[inline]
-    pub fn watch_client(&mut self) -> WatchClient {
+    pub fn watch_client(&self) -> WatchClient {
         self.etcd_client.watch_client()
     }
+
+    /// Gets an lock client.
```
Comment (Phoenix500526) [inline xline/src/client/mod.rs:270]: Same above

File: xline/src/client/mod.rs [hunk 10]
```diff
@@ -238,15 +249,33 @@ impl Client {
...
+    #[inline]
+    pub fn lock_client(&self) -> LockClient {
+        self.etcd_client.lock_client()
+    }
+
+    /// Gets an lease client.
```
Comment (Phoenix500526) [inline xline/src/client/mod.rs:276]: Same above

File: xline/src/server/kv_server.rs (hunk 2 above)
Comment (themanforfree) [reply xline/src/server/kv_server.rs:438]: In the future we may let the client decide whether to use fast path，so I think current logic is better for this.

File: xline/src/server/xline_server.rs (hunk 3 above)
Comment (Phoenix500526) [reply xline/src/server/xline_server.rs:56]: Is it a good idea to add this dependency to `client` here? I mean, the xline client is only used in `test` and `benchmark` before implementing the `Lock` feature.

File: xline/src/client/mod.rs (hunk 4 above)
Comment (themanforfree) [reply xline/src/client/mod.rs:254]: This is just a temporary client implementation, it will be refactor in the future.

**Review by Phoenix500526 — APPROVED** [review]

**Review by markcty — APPROVED** [review]

**Pull request merged**